app/models/claim.rb

```diff
@@ -246,6 +246,27 @@ def school
     eligibility&.current_school
   end
 
+  def inconsistent_claims
```

Comment:
There's quite a lot of logic in here, and I found it tricky to follow, and I struggled to see the rationale for some things that were going on (for example the call to `attributes.delete(:national_insurance_number)`). I'm wondering if it might be better to wrap this in a class (something like `InconsistentClaimFinder`). This finder could also have a method to return the inconsistent attributes too, to avoid having to pass the inconsistent claims in again in the ClaimsController.

Reply:
So, I think the best thing is if I explain what the code is _meant_ to be doing, and then I can understand whether you think that my approach is unnecessarily complicated, or whether I just need to figure out how to make the current implementation more understandable.

The aim of the method is the following: given a claim `c`, we want to find all the claims which:
- are payrollable;
- have the same National Insurance number as our claim;
- have a different value to `c` for at least one of the attributes in `Payment::PERSONAL_DETAILS_ATTRIBUTES_FORBIDDING_DISCREPANCIES`.

The first thing I had to decide was: given two claims, what's the best way of comparing their values for these attributes?
- I initially tried writing code like `.where.not(student_loan_plan_type: read_attribute(:student_loan_plan))`, but this didn't work because `read_attribute(:student_loan_plan)` returns a string `"plan_1"` but the database needs the value to be specified as the integer enum value. So instead of trying to figure out how to correctly re-serialize each attribute for insertion into the equality query, I decided to let the database to directly compare the two records. This is why I join the `claims` table to itself, so that I can compare the other claims to claim `c` (represented by the `this_claim` in the `JOIN`). I'm not very happy about the join, though.
- I wanted a comparision which treated `NULL` as "equal" to `NULL` (for the case of building society roll number, which is nullable), hence using `IS DISTINCT FROM` instead of `=`.

What do you think?

(And as for putting it into a separate class, happy to try that but the need may fall away if it turns out that the method can be hugely simplified.)

Reply:
Ah, and the `delete(:national_insurance_number)` is because when building up the list of inconsistent attributes to look for, we know that there's no point looking for National Insurance Number mismatch, because we use NI number to pick the records that we want to compare with, so there will never be a mismatch.

Reply:
I reckon it's looking a lot better now, but I still think it'd be best to break out into a seperate class, partly because there's quite a lot of logic there, and also partly because the Claim class is getting rather overloaded. How about something like this?

```ruby
  class ClaimsPreventingPaymentFinder
    attr_reader :claim

    def initialize(claim)
      @claim = claim
    end

    def claims
      @claims ||= begin
        Claim.payrollable
          .joins("INNER JOIN claims AS this_claim ON claims.national_insurance_number = this_claim.national_insurance_number \
              AND (#{distinct_attributes_predicate})")
          .where("this_claim.id = ?", claim.id)
      end
    end

    def attributes_preventing_combined_payment
      Payment::PERSONAL_DETAILS_ATTRIBUTES_FORBIDDING_DISCREPANCIES.select do |attribute|
        claims.any? { |claim| claim.read_attribute(attribute) != read_attribute(attribute) }
      end
    end

    private

    def distinct_attributes_predicate
      distinct_attributes_predicate = attributes_to_search_for.map { |attribute|
        "claims.#{attribute} IS DISTINCT FROM this_claim.#{attribute}"
      }.join(" OR ")
    end

    def attributes_to_search_for
      attributes = Payment::PERSONAL_DETAILS_ATTRIBUTES_FORBIDDING_DISCREPANCIES.dup
      attributes.delete(:national_insurance_number)
      attributes
    end
  end
```

Reply:
Moved it to a separate class, as suggested.

So, after getting lost in a failing test rabbit hole that turned out to be because factory_bot was silently failing to set up some associations, and thinking that this was because of the wacky SQL (which it turned out it wasn't), I took your advice and moved some of this comparison out of the SQL and into the code. I've added some justification to the commit message.